### PR TITLE
Fix check for outdated Rust library

### PR DESCRIPTION
**Pull request by erikjohnston**
This failed when install with poetry, so let's properly try and detect what's going on.

### CODE REVIEW

File: changelog.d/17861.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix detection when the built Rust library was outdated when using source installations.
```
Comment (MadLittleMods) [inline changelog.d/17861.bugfix:1]: > This failed when install with poetry, so let's properly try and detect what's going on.

What's the testing strategy for this?

 1. Create a new directory
 1. `poetry new synapse-pr-17861-test`
 1. `cd synapse-pr-17861-test`
 1. `poetry add synapse` (but install from this branch)
 1. ...

Comment (erikjohnston) [reply changelog.d/17861.bugfix:1]: This is for source check outs, so run `poetry install -E "all test"` in a checkout of this branch. Then `python -c "import synapse"` should work, while if you edit a rust source file it should then fail.

Comment (MadLittleMods) [reply changelog.d/17861.bugfix:1]: Looks good :white_check_mark: 

Test what happens before this change:

 1. `git clone git@github.com:element-hq/synapse.git synapse-pr-17861-test`
 1. `cd synapse-pr-17861-test`
 1. `poetry install --extras all`
 1. `poetry run python -c "import synapse"` (works as expected)
 1. Make an edit to one of the rust files like `rust/src/http.rs`
 1. `poetry run python -c "import synapse"` (works when it shouldn't :x:)

Test how this PR fixes things:
 
 1. `git checkout erikj/fix_outdated_rust_detection`
 1. Actually complains about dirty Rust source files (worked as expected)
    ```sh
    $ poetry run python -c "import synapse"
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/home/eric/Downloads/synapse-pr-17861-test2/synapse/__init__.py", line 107, in <module>
        check_rust_lib_up_to_date()
      File "/home/eric/Downloads/synapse-pr-17861-test2/synapse/util/rust.py", line 51, in check_rust_lib_up_to_date
        raise Exception("Rust module outdated. Please rebuild using `poetry install`")
    Exception: Rust module outdated. Please rebuild using `poetry install`
    ```
 1. Re-install as instructed to rebuild rust: `poetry install --extras all`
 1. `poetry run python -c "import synapse"` (works as expected :white_check_mark:)

**Review by MadLittleMods — APPROVED** [review]

**Comment by erikjohnston** [comment]
Thank you!

**Pull request merged**